### ISSUE TITLE
Can't download files with ?format= (HTTP 302 / CORS issue)

### ISSUE THREAD

**Issue by bitflower** [microsoftgraph/msgraph-sdk-javascript#196]
# Bug Report

## Prerequisites

-   [x] Can you reproduce the problem?
-   [x] Are you running the latest version?
-   [x] Are you reporting to the correct repository?
-   [x] Did you perform a cursory search?

## Description

Thanks a lot for resolving my issue https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/193 recently. I could move on with some requirements. I'm facing a new issue right now when downloading files in PDF format. It should be supported as per the docs here:

I'm getting the error below (in Chrome, latest version).

As described here I should use the returned URL to download the file. However for security reasons this error can not be tracked in Javascript inside browsers. Therefore the proposed solution will never work: https://docs.microsoft.com/en-us/graph/api/driveitem-get-content-format?view=graph-rest-1.0&tabs=javascript
https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_get_content_format?view=odsp-graph-online

I guess the right solution would be to receive a HTTP 200 with the download url which we can then fetch.

**Console Errors:** 
```
Access to fetch at 'https://westeurope1.mediap.svc.ms/transform/pdf?provider=ODC&inputFormat=xlsx&docID=https%3A%2F%2Fs2s-api.onedrive.com%2Fv1.0%2Fdrives%28%2741f5edfbb62f164a%27%29%2Fitems%28%2741f5edfbb62f164a%21988%27%29&ci=8abbcccf-512f-44ba-9c9b-a415cb267110%7C1277658449&si=ODC&cid=41f5edfbb62f164a&tempauth=<TEMP_AUTH>' (redirected from 'https://graph.microsoft.com/v1.0/me/drive/items/41F5EDFBB62F164A!988/content?format=pdf') from origin 'null' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Screenshots:** [If applicable, add screenshots to help explain your problem]

## Steps to Reproduce

You can find the repo here: https://github.com/bitflower/stencil-ms-graph

**Expected behavior:** 
Download in PDF format is triggered.

**Actual behavior:**

**Comment by muthurathinam** [comment]
@bitflower  Refer this comment: https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/109#issuecomment-430093219

**Comment by bitflower** [comment]
Hi, thanks for the info. I‘m aware of this but was thinking it doesn‘t support PDF file format conversion? Say from excel to PDF...?

**Comment by muthurathinam** [comment]
@ificator Can you help with correct API to achieve this ?

**Comment by ificator** [comment]
What request headers are being provided. That endpoint does have some preflight issues iirc, but realistically the GET request that returns the `PDF` data should be considered "simple" and not require preflight anyway.

**Comment by bitflower** [comment]
Hi,
this is all I can provide. It what happens from when I call 
```
    const format: string = 'pdf';
    const pdfUrl: string = `/me/drive/items/${
      this.rootFolders[0].id
    }/content?format=${format}`;
    console.log('pdfUrl', pdfUrl);

    try {
      let res = await this.graphClient.api(pdfUrl).get();
      console.log('BLUB', res);
    } catch (error) {
      // Extract PDF URL
      console.log('GOT IT!', error);
      throw error;
    }
  }
```

<img width="1436" alt="1" src="https://user-images.githubusercontent.com/6666636/60297047-ac008400-9927-11e9-986e-c210338cc381.png">
<img width="1739" alt="2" src="https://user-images.githubusercontent.com/6666636/60297045-ac008400-9927-11e9-870a-e0d1ed3ee4c3.png">
<img width="1887" alt="3" src="https://user-images.githubusercontent.com/6666636/60297044-ac008400-9927-11e9-9a6d-317a6daf22bf.png">
<img width="1895" alt="4" src="https://user-images.githubusercontent.com/6666636/60297043-ac008400-9927-11e9-9ffd-de3bcbeb0fb6.png">
<img width="1914" alt="5" src="https://user-images.githubusercontent.com/6666636/60297042-ac008400-9927-11e9-98af-56f7f6226ca0.png">

**Comment by ificator** [comment]
So it looks like there are three problematic headers:
1. authorization
2. client-request-id
3. sdkversion

It sounds like the SDK is following the redirect itself, and it dutifully attaches its usual headers when it does so. It's this behavior that's triggering CORS.

Now my knowledge of the SDK is limited, but ideally for redirects related to `/content` scenarios (including the use of `format`) the SDK should follow the direct but NOT add any of its usual headers so that the preflight is avoided.

**Comment by bitflower** [comment]
Thanks for investigating.

Let me know how I can help.

**Comment by bitflower** [comment]
Hi guys, did you have the chance to investigate a little deeper here? Thanks!

**Comment by muthurathinam** [comment]
SDK is just following the redirect it did not add any additional headers on redirected request.
@bitflower I have tried setting up this in my local server it works fine for me now. can you verify once again. if its unsuccessful try setting the `redirect option` like,
```js
let res = await this.graphClient
	        .api(pdfUrl)
		.option("redirect", "follow")
		.get();
```

**Comment by bitflower** [comment]
Hi @muthurathinam, thanks for the hint. Unfortunately I get the same error:
<img width="1717" alt="Bildschirmfoto 2019-07-11 um 21 05 31" src="https://user-images.githubusercontent.com/6666636/61077951-c2c2d280-a41f-11e9-945b-6e71e7c7462e.png">

I've used this code:
```
    const format: string = 'pdf';
    const pdfUrl: string = `/me/drive/items/${
      this.rootFolders[0].id
    }/content?format=${format}`;
    console.log('pdfUrl', pdfUrl);

    try {
      let res = await this.graphClient
        .api(pdfUrl)
        .option('redirect', 'follow')
        .get();

      console.log('BLUB', res);

      // this.rootFolders = res.value;
    } catch (error) {
      // Extract PDF URL
      console.log('GOT IT!', error);
      throw error;
    }
```

Have you tried my repo with an app id of yours?

Any other things I can try?

**Comment by bitflower** [comment]
Hi guys, did you have the chance to check my repo?

**Comment by Cutaiar** [comment]
Any update on this? I am facing a similar issue even without using `format` -- even just trying to access `/me/drive/items/${itemId}/content` results in a CORS issue.

**Comment by nikithauc** [comment]
Duplicate of #16. 

Workaround for this problem - https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/316#issuecomment-718264064

**Comment by nikithauc** [comment]
Closing this as this is a known issue. Shall update if the CORS issue gets resolved.
